src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -4,56 +4,100 @@ package freechips.rocketchip.diplomacy
 
 import Chisel._
 import chisel3.{RawModule, MultiIOModule, withClockAndReset}
-import chisel3.internal.sourceinfo.{SourceInfo, SourceLine, UnlocatableSourceInfo}
+import chisel3.internal.sourceinfo.{SourceInfo, UnlocatableSourceInfo}
 import freechips.rocketchip.config.Parameters
-import scala.collection.immutable.{SortedMap,ListMap}
+import scala.collection.immutable.{SortedMap, ListMap}
 import scala.util.matching._
 
+/** [[LazyModule]] uses Scala's lazy evaluation property to 
+  *  split hardware generation into two phases.
+  * - Phase 1 (non-lazy) states hardware intent and constraints:
+  * -- [[LazyModule]] and [[BaseNode]] instantiation.
+  * -- [[BaseNode]] binding.
+  * - Phase 2 (lazy) generates hardware:
+  * -- Parameter negotiation.
+  * -- [[AutoBundle]] resolution.
+  * -- [[LazyModuleImpLike]] module generation.
```

Comment:
I agree that it is only the `Data` that makes this at all chisel-y. But, that is the current implementation, which we are documenting. If we want to change what this is, then we could change the documentation at the same time. Do you agree?

Reply:
My worry is that by defining it specifically in terms of hardware generation, it obfuscates the essence of the implementation and makes it more difficult to understand the code, since the vast majority of it is unrelated to Chisel or hardware.

I think this is less obvious in `LazyModule.scala`, which is more directly tied to hardware than other parts of Diplomacy. If you instead take a look at [`Nodes.scala`](https://github.com/chipsalliance/rocket-chip/blob/f31e21b5b6eb04f27320ff9c8fab3f1661f5b1f1/src/main/scala/diplomacy/Nodes.scala), you'll see that there is almost nothing related to hardware or Chisel there. The only thing in that file related to hardware is the `Data` type.

Reply:
I was thinking about htis more. This file (LazyModule.scala) really is teh place where we are taking the abstract diplomacy "thing" from `Nodes` and turning it into hardware. Maybe we should make that distinction explicit in this description?

Reply:
I suggested a change for this section